In [12]:
import pandas as pd

df = pd.read_csv('survey_results_public.csv', encoding='utf-8', low_memory=False)
df1 = pd.read_csv('survey_results_schema.csv', encoding='utf-8')

pd.set_option('display.max_columns', None)

ParserError: Error tokenizing data. C error: Expected 172 fields in line 864, saw 322


In [ ]:
df

In [ ]:
df1

**Висновки:**

**Успішне завантаження:** Робоче середовище Google Colab налаштовано, дані за 2025 рік успішно зчитані у змінні df та df1.

**Обсяг та структура:** Датасет містить **49 191** рядків та **172** колонки. Велика кількість текстових даних object вказує на потребу в подальшій перевірці та очищенні для аналізу числових показників.

**Технічна готовність:** Завдяки налаштуванню pd.set_option, текст курсивом текст курсивом забезпечено зручний доступ до всіх параметрів опитування. Етап підготовки завершено, дані готові до обробки.

In [ ]:
total_respondents = df['ResponseId'].nunique()

print(f"Загальна кількість респондентів: {total_respondents}")

**Висновок:**

**Загальна кількість:** Опитування охоплює **49 191** унікального респондента.

**Метод підрахунку:** Використання .nunique() за колонкою ResponseId підтвердило відсутність дублікатів у даних.

**Готовність:** Отримана вибірка є репрезентативною для подальшого аналізу трендів ІТ-індустрії.

In [ ]:
questions = set(df1['qname'].unique()) & set(df.columns)

full_responses_count = df[list(questions)].dropna().shape[0]

print(f"Кількість респондентів, які відповіли на всі запитання: {full_responses_count}")

**Висновок:**

**Результат аналізу:** Кількість респондентів, які надали відповіді на всі запитання без винятку, становить **0**.

**Причина:** Опитування має складну структуру з багатьма специфічними та необов'язковими блоками. Наявність пропусків є природною, оскільки респонденти відповідають лише на ті питання, що стосуються їхнього досвіду.

**Технічний підсумок:** Використання перетину множин гарантувало, що аналіз повноти відповідей проводився виключно за тими колонками, які офіційно зафіксовані у схемі опитування qname.

In [ ]:
print(f'Міри центральної тенденції :')

print(f'Mean: {round(df.WorkExp.mean(), 1)}')

print(f'Median: {df.WorkExp.median()}')

print(f'Mode: {df.WorkExp.mode()[0]}')

**Висновок:**

**Центральна тенденція:** Обчислено основні статистичні показники для колонки WorkExp (досвід роботи респондентів).

**Результати:** Середній стаж респондентів становить **13.4** року, тоді як медіанне значення **10.0** років. Найпопулярнішим показником досвіду (мода) є **10.0** років.

**Аналіз:** Різниця між середнім - 13.4 та медіаною - 10.0 вказує на наявність у вибірці досвідчених спеціалістів, які зміщують середнє значення вгору. Проте типовий респондент має близько **10** років досвіду.

In [ ]:
remote_count = df[df['RemoteWork'] == 'Remote'].shape[0]

print(f"Кількість респондентів, які працюють віддалено: {remote_count}")

**Висновок:**

**Результат:** Кількість респондентів, які працюють віддалено — **10931**.

**Технічний підсумок:** Розрахунок виконано шляхом фільтрації колонки RemoteWork за значенням "Remote" із застосуванням .shape[0].

**Контекст:** Отримане число дозволяє оцінити частку дистанційної праці у загальній вибірці та порівняти її з іншими форматами зайнятості.

In [ ]:
df['worked_with_python'] = df['LanguageHaveWorkedWith'].str.contains('python', case=False, na=False)

python_users_count = df['worked_with_python'].sum()

total_responses = df['LanguageHaveWorkedWith'].dropna().count()
python_percentage = (python_users_count / total_responses) * 100

print(f"Відсоток респондентів: {round(python_percentage, 1)}%")

**Висновок:**

**Результат аналізу:** Частка респондентів, які використовують мову Python, становить **58.3%**.

**Технічний підсумок:** Для аналізу було створено допоміжний булевий стовпець worked_with_python. Використання методу .str.contains() із параметром case=False забезпечило точне виявлення Python у рядках із множинними значеннями мов.

**Аналіз:** Понад половина опитаних фахівців працюють з Python, що робить цю мову ключовим інструментом у вибірці. Це пояснює високий попит на Python-розробників та аналітиків даних.

In [ ]:
df['learned_with_online_courses'] = df['LearnCode'].str.contains('Online Courses', case=False, na=False)

online_learners_count = df['learned_with_online_courses'].sum()

print(f"Кількість респондентів, які навчалися через онлайн-курси: {online_learners_count}")

**Висновок:**

**Результат аналізу:** Кількість респондентів, які зазначили онлайн-курси як один із методів навчання — **10973**.

**Технічний підсумок:** Для ідентифікації відповідних записів створено булеву маску через метод .str.contains. Це дозволило коректно опрацювати множинні відповіді в колонці LearnCode, де онлайн-курси часто згадуються разом з іншими методами.

**Аналіз:** Отримані дані підтверджують, що онлайн-освіта є одним із найпопулярніших способів здобуття знань в ІТ, що свідчить про гнучкість та доступність сучасних освітніх інструментів.

In [ ]:
python_devs_with_salary = df[(df['worked_with_python'] == True) & (df['ConvertedCompYearly'].notna())]

py_progers_by_country = python_devs_with_salary.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median']).round(1)

py_progers_by_country.columns = ['Середня компенсація', 'Медіанна компенсація']

py_progers_by_country = py_progers_by_country.sort_values(by='Медіанна компенсація', ascending=False)

py_progers_by_country.head(20)

**Висновок:**

**Результат аналізу:** Сформовано рейтинг країн за рівнем річної компенсації Python-розробників. Лідерами за медіанним доходом у цій вибірці стали:

**Оман ($390,135)**

**Андорра ($226,103)**

**США ($150,000)**

**Ізраїль ($142,594)**

**Технічний підсумок:** Для розрахунку використано комбінацію методів .groupby('Country') та .agg(['mean', 'median']). Попередня фільтрація забезпечила точність обчислень, а результати впорядковані за спаданням медіани для нівелювання впливу одиничних надвисоких зарплат.

**Аналітичні спостереження:**

1. У багатьох країнах середня зарплата вища за медіану (як-от у США). Це підтверджує наявність "викидів" — поодиноких фахівців із надвисокими доходами, які штучно завищують середній показник.
2. Повна ідентичність середнього та медіани для Оману та Андорри вказує на малу кількість респондентів у цих локаціях. Показник США залишається найбільш репрезентативним еталоном через великий обсяг даних.
3. Високі позиції Швейцарії, Ізраїлю та категорії Nomadic підтверджують, що ринок Python-розробки є високооплачуваним у всьому світі, незалежно від географічної прив'язки.

In [ ]:
top_5_edu = df[['ResponseId', 'EdLevel', 'ConvertedCompYearly']].sort_values(by='ConvertedCompYearly', ascending=False).reset_index(drop=True).head(5)

top_5_edu

**Висновок:**

**Результат аналізу:** Сформовано перелік 5 респондентів з найвищою річною компенсацією. Результати демонструють аномально високі рівні доходу, які значно перевищують ринкові стандарти:

**Associate degree: $50,000,000 (ResponseId: 34268)**

**Master’s degree: $33,552,715 (ResponseId: 28701)**

**Associate degree: $18,387,548 (ResponseId: 43144)**

**Технічний підсумок:**

1. Виконано вибірку стовпців: ResponseId, EdLevel та ConvertedCompYearly.
2. Застосовано сортування за спаданням (ascending=False).
3. Використано .reset_index(drop=True) для чистої нумерації рейтингу (0–4) та .head(5) для обмеження вибірки.

**Аналітичні спостереження:**

1. Виявлені суми (до $50 млн) є очевидними екстремальними викидами. Такі показники зазвичай свідчать про помилки респондентів при заповненні анкети — наприклад, введення місячного доходу замість річного (що при автоматичній конвертації дає мільйонні суми) або помилку у виборі валюти.
2. Дані показують, що в межах цього датасету найвищі, хоч і аномальні зарплати, не мають прямої залежності від наукового ступеня (наприклад, Associate degree випереджає Master’s). Це ще раз підкреслює, що в ІТ практичний досвід та конкретні навички часто важать більше, ніж формальна освіта.
3. Наявність таких записів підтверджує, що для аналізу реального ринку слід орієнтуватися виключно на медіану, оскільки середнє арифметичне через подібні випадки стає абсолютно нерепрезентативним.

In [ ]:
age_python_stats = df.assign(is_python=df['LanguageHaveWorkedWith'].str.contains('Python', na=False)).groupby('Age')['is_python'].mean().mul(100).round(2).reset_index().rename(columns={'Age': 'Вікова категорія', 'is_python': 'Відсоток Python-розробників (%)'}).sort_values(by='Відсоток Python-розробників (%)', ascending=False)

age_python_stats

**Висновок:**

**Результат аналізу:** Сформовано рейтинг популярності **Python** серед різних вікових груп. Лідером є категорія **18-24 years old (40.0%)**, а найменший відсоток спостерігається у групі **Prefer not to say (31.22%)**.

**Технічний підсумок:**

1. Створено індикатор володіння мовою через пошук підрядка "Python" у колонці технологій.
2. Проведено групування за віком з розрахунком середнього значення (частки) позитивних відповідей.
3. Дані трансформовано у відсотки, заокруглено та відсортовано за спаданням для наочності рейтингу.

**Аналітичні спостереження:**

1. Найвищий рівень використання **Python** серед молоді **(40.0%)** підтверджує статус мови як основної для навчання та старту кар'єри в IT.
2. Цікаво, що групи **від 25 до 54 років** демонструють дуже стабільний і близький показник **(в районі 36-38%)**, що свідчить про універсальність мови для досвідчених спеціалістів.
3. Навіть у найстаршій категорії **(65 years or older)** кожен третій розробник **(31.63%)** використовує Python, що підкреслює його актуальність незалежно від віку.

In [ ]:
res = df[(df['ConvertedCompYearly'] > df['ConvertedCompYearly'].quantile(0.75)) & (df['RemoteWork'] == 'Remote')]['Industry'].value_counts().reset_index().rename(columns={'Industry': 'Індустрія', 'count': 'Кількість фахівців'})

industries_top_remote = pd.concat([res[res['Індустрія'] != 'Other:'], res[res['Індустрія'] == 'Other:']], ignore_index=True)

industries_top_remote

**Висновок:**

**Результат аналізу:** Сформовано рейтинг індустрій для найбільш оплачуваних віддалених працівників (75-й перцентиль). Абсолютним лідером є **Software Development (1186)**, а друге та третє місця посідають **Fintech (190)** та **Healthcare (188)**. Категорію **Other (176)** переміщено в кінець списку для покращення візуального сприйняття основних галузей.

**Технічний підсумок:**

1. Розраховано поріг 75-го перцентиля за допомогою .quantile(0.75) та відфільтровано респондентів за критеріями доходу та формату роботи "Remote".
2. Виконано підрахунок частоти індустрій через .value_counts().
3. Застосовано метод pd.concat для кастомного сортування, що дозволило відокремити специфічні індустрії від загальних відповідей та оновити індексацію через ignore_index=True.

**Аналітичні спостереження:**

1. **Software Development** охоплює левову частку високооплачуваних віддалених позицій **(1186 фахівців)**, що підтверджує цифрову зрілість цієї галузі.
2. **Fintech** та **Healthcare** демонструють майже ідентичні показники (**190** та **188** відповідно), що вказує на однаково високу концентрацію капіталу та готовність цих індустрій платити за віддалену експертизу.
3. Галузь **Internet** та **Telecom** також входить у **топ-4** **(138 фахівців)**, замикаючи групу основних лідерів ринку.

**Загальні підсумки дослідження:**

**Дані:** Опрацьовано повний датасет (49 191 запис). Очищення від NaN забезпечило точність розрахунків.

**Зарплати:** Виявлено аномалії до $50 млн, тому для звіту використано медіану.

**Python:** Найпопулярніший серед молоді 18–24 років (40.0%).

**Індустрії:** Лідери віддаленої роботи — Software Development (1186), Fintech (190) та Healthcare (188).

**Висновок:** Рівень доходу прямо залежить від стажу (медіана 10 років) та спеціалізації у високонавантажених секторах.

**Ключовий інсайт:** Аналіз демонструє, що поєднання навичок Python-розробки з експертизою у фінансових чи медичних технологіях відкриває доступ до найбільш високооплачуваних сегментів ринку. Висока медіана досвіду (**10 років**) серед респондентів підтверджує, що стабільність доходу в індустрії безпосередньо корелює з практичним стажем.
